File: tools/excel/src/main/java/top/bettercode/summer/tools/excel/ExcelRangeCell.java
package top.bettercode.summer.tools.excel;

/**
 * @author Peter Wu
 */
public class ExcelRangeCell<T> extends ExcelCell<T> {

  private final int lastRangeTop;
  private final int lastRangeBottom;
  private final Object preCellValue;
  private final boolean needSetValue;
  private final boolean needRange;
  private final boolean newRange;

  public ExcelRangeCell(int row, int column, int index, int firstRow, boolean lastRow,
      boolean newRange, int lastRangeTop,
      ExcelField<T, ?> excelField, T preEntity, T entity) {
    super(row, column, lastRow, excelField.isMerge() ? index : row - firstRow + 1, index % 2 == 0,
        excelField, entity);

    this.newRange = newRange;
    this.lastRangeTop = lastRangeTop;

    this.lastRangeBottom = newRange && index > 1 ? row - 1 : row;
    boolean mergeLastRange = (newRange || lastRow) && lastRangeBottom > lastRangeTop;
    this.needRange = mergeLastRange && getExcelField().isMerge();
    this.needSetValue = !getExcelField().isMerge() || newRange;

    if (!excelField.isIndexColumn() && newRange && preEntity != null) {
      preCellValue = excelField.toCellValue(preEntity);
    } else {
      preCellValue = null;
    }
  }

  //--------------------------------------------


  public boolean newRange() {
    return newRange;
  }

  public boolean needSetValue() {
    return needSetValue;
  }

  public boolean needRange() {
    return needRange;
  }

  //--------------------------------------------

  public Object getPreCellValue() {
    return preCellValue;
  }

  public int getLastRangeTop() {
    return lastRangeTop;
  }

  public int getLastRangeBottom() {
    return lastRangeBottom;
  }

}

File: tools/excel/src/main/kotlin/top/bettercode/summer/tools/excel/ExcelRangeCell.kt
package top.bettercode.summer.tools.excel

/**
 * @author Peter Wu
 */
class ExcelRangeCell<T>(row: Int, column: Int, index: Int, firstRow: Int, lastRow: Boolean,
                        val newRange: Boolean, val lastRangeTop: Int,
                        excelField: ExcelField<T, *>, preEntity: T?, entity: T) : ExcelCell<T>(row, column, lastRow, if (excelField.isMerge) index else row - firstRow + 1, index % 2 == 0,
        excelField, entity) {
    val lastRangeBottom: Int

    //--------------------------------------------
    var preCellValue: Any? = null
    private val needSetValue: Boolean
    val needRange: Boolean

    init {
        lastRangeBottom = if (newRange && index > 1) row - 1 else row
        val mergeLastRange = (newRange || lastRow) && lastRangeBottom > lastRangeTop
        needRange = mergeLastRange && excelField.isMerge
        needSetValue = !excelField.isMerge || newRange
        preCellValue = if (!excelField.isIndexColumn && newRange && preEntity != null) {
            excelField.toCellValue(preEntity)
        } else {
            null
        }
    }

    //--------------------------------------------

    override fun needSetValue(): Boolean {
        return needSetValue
    }

}
